include/dali/core/mm/memory.h

```diff
@@ -0,0 +1,153 @@
+ * buffer, but does not call destructors of the objects.
+ *
+ * @param count     Number of objects for which the storage should suffice.
+ * @tparam T        Type of the object for which the storage is allocated.
+ * @tparam kind     The kind of requested memory.
+ * @tparam Context  The execution context in which the memory will be available.
+ */
+template <typename T, memory_kind kind>
+auto alloc_raw(size_t count) {
+  return alloc_raw<T>(GetDefaultResource<kind>(), count);
+}
+
+/**
+ * @brief Allocates uninitialized storage for `count` objects of type `T`
+ *        with stream-ordered semantics
+ *
+ * This function allocates memory with stream-ordered semantics. The `alloc_stream` denotes
+ * the stream on which the memory can be safely used without additional synchronization.
+ * Use the value `host_sync` if the memory needs to be accessible on all streams or on host as
+ * soon as the function returns.
+ * `dealloc_stream` denotest the stream for deallocation. Stream-ordered semantics guarantee,
+ * that if there's still some work pending on `dealloc_stream`, it will finish before the memory
+ * returned by this function is freed. Use `host_sync` for host-synchronous execution.
+ *
+ * This function allocates raw memory with suitable size and alignment to accommodate `count`
+ * object of type `T`. The memory is obtained from memory resource `mr` with stream semantics.
+ * The size of memory is at least sizeof(T) * count and the alignment is at least alignof(T).
+ * The return value is a unique pointer with a deleter that will safely dispose of the allocated
+ * buffer, but does not call destructors of the objects.
+ *
+ * @param mr              Memory resources to allocate the memory from
+ * @param count           Number of objects for which the storage should suffice
+ * @param alloc_stream    The CUDA stream on which the memory is immediately usable
+ * @param dealloc_stream  The CUDA stream which is guaranteed to finish all work scheduled
+ *                        before the deallocation of the memory.
+ * @tparam T              Type of the object for which the storage is allocated.
+ * @tparam kind           The kind of requested memory.
+ */
+template <typename T, memory_kind kind>
+mr_stream_unique_ptr<T, kind> alloc_raw_async(async_memory_resource<kind> *mr,
+                                              size_t count,
+                                              cudaStream_t alloc_stream,
+                                              cudaStream_t dealloc_stream) {
+  size_t bytes = sizeof(T) * count;
+  size_t alignment = alignof(T);
+  void *mem = alloc_stream == host_sync
+    ? mr->allocate(bytes, alignment)
+    : mr->allocate_async(bytes, alignment, alloc_stream);
+  return mr_stream_unique_ptr<T, kind>(static_cast<T*>(mem),
+      mr_stream_deallocator<kind>{mr, bytes, alignment, dealloc_stream});
+}
+
+template <typename T, memory_kind kind>
```

Comment:
Does it deserve a separate doxygen doc as `alloc_raw` do?

Reply:
Done.